src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+                    let message = Message::PoolRequest(share_difficulty, Data::Object(block_template.clone()));
+                    if let Err(error) = self.peers_router.send(PeersRequest::MessageSend(peer_ip, message)).await {
+                        warn!("[PoolRequest] {}", error);
+                    }
+                } else {
+                    warn!("[PoolRegister] No current block template exists");
+                }
+            }
+            OperatorRequest::PoolResponse(peer_ip, block, prover_address) => {
+                if let Some(block_template) = &*self.block_template.read().await {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block.height() {
+                        warn!("[PoolResponse] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Retrieve the coinbase transaction records.
+                    let coinbase_records = match block.to_coinbase_transaction() {
+                        Ok(transaction) => {
+                            // Ensure the owner of the coinbase transaction in the block is the operator address.
+                            let coinbase_records: Vec<Record<N>> =
+                                transaction.to_records().filter(|r| Some(r.owner()) == self.address).collect();
+                            if coinbase_records.len() == 0 {
+                                warn!("[ProposedBlock] Peer {} sent a candidate block with an incorrect owner.", peer_ip);
+                                return;
+                            }
+                            coinbase_records
+                        }
+                        Err(error) => {
+                            warn!("[PoolResponse] {}", error);
+                            return;
+                        }
+                    };
+
+                    // Ensure the block contains a difficulty that is at least the share difficulty.
+                    let proof_difficulty = {
+                        let proof_bytes = match block.header().proof() {
+                            Some(proof) => match proof.to_bytes_le() {
+                                Ok(bytes) => bytes,
+                                Err(error) => {
+                                    warn!("[PoolResponse] {}", error);
+                                    return;
+                                }
+                            },
+                            None => {
+                                warn!("[PoolResponse] Peer {} sent a candidate block with a missing proof.", peer_ip);
+                                return;
+                            }
+                        };
+
+                        let proof_difficulty = sha256d_to_u64(&proof_bytes);
+                        let share_difficulty = {
+                            let mut provers = self.provers.write().await;
```

Comment:
is it much more likely that the new prover will be inserted rather than looked up? otherwise I'd do a separate `read` and `write` in order to limit the number of writes